observability/config.go

```diff
@@ -0,0 +1,97 @@
+/*
+ *
+ * Copyright 2022 gRPC authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ */
+
+package observability
+
+import (
+	"context"
+	"encoding/json"
+	"os"
+
+	coptlogging "cloud.google.com/go/logging"
+	"golang.org/x/oauth2/google"
+	configpb "google.golang.org/grpc/observability/internal/config"
+	"google.golang.org/protobuf/encoding/protojson"
+)
+
+const (
+	envKeyObservabilityConfig = "GRPC_OBSERVABILITY_CONFIG"
+)
+
+// gcpDefaultCredentials is the JSON loading struct used to get project id.
+type gcpDefaultCredentials struct {
+	QuotaProjectID string `json:"quota_project_id"`
+}
+
+// fetchDefaultProjectID fetches the default GCP project id from environment.
+func fetchDefaultProjectID(ctx context.Context) string {
```

Comment:
Thanks @codyoss.  Doesn't client libraries need very similar code as this for its own purposes?  Or do you require the user to always set the project ID explicitly?  Or do you only care about the project ID via the credentials?

If similar code does exist, can you point us at it?  It seems _some_ library should have this logic, as opposed to requiring duplicating it all across all clients / shelling out to `gcloud`.  `oauth2` doesn't seem like the right place to me, either.  Is this a function that could be exported from some public cloud client libraries package?

Reply:
Maybe it's just me, in my local environment, I use `gcloud auth default-application login` to setup ADC. But my project ID is embeded in its JSON field:

```json
{"ProjectID":"","TokenSource":{},"JSON":"ewogICJjbGllbnRfaWQiOiAiNzY0MDg2MDUxODUwLTZxcjRwN..."
```

Unpacking the JSON field, it will be:

```json
{"quota_project_id":"grpc-testing"}
```

Is this scenario handled by `transport.Creds`?

Reply:
quota project is different than project. Quota project == billing project. With user credentials I don't think there is a field for just project id like there is for service accounts, I am not sure if we should use quota project for that. I think this ambiguity reenforces my thought that we need a consistent cross lang design for this.

Reply:
Maybe we should just go ahead and make the project ID a required field in the observability config for now?

Reply:
@dfawley Can we still check those 2 common places: 1. `GOOGLE_CLOUD_PROJECT`; 2. Default creds' ProjectID field. This way aligns with SDK's implementation. It should cover the majority of use cases in GCP. For corner cases, we will log that no-project ID found.

Reply:
So you'll remove the quota project ID part, then?  That seems okay.  But then let's make sure we are logging what project ID we found and how we found it.